drivers: ksan: add new APIs for keyboard matrix scan task

ChiHuaL opened the pull request:
Two types of keyboard scan drivers are supported in the codebase:
        - Remote type: the kscan hardware/algorithm exists in a remote chip.
          The SoC gets the key code and press/release event via a peripheral
           bus like I2C.
        - Embedded type: the kscan hardware/algorithm exists in the SoC.
For the second case, the existing kscan driver generally has two major
tasks to do:
        - control/receive the hardware signals, such as drive a column or
          read a row.
        - implement the kscan algorithm task.
    
However, an application may want to implement its scan algorithm by
driving the low-level hardware directly.
This PR extends the APIs to achieve it because the kscan matrix
needn't  adhere to the driver anymore.

This PR also adds
    - npcx kscan driver support.
    - a general kscan matrix task (basically, most the codes extracted from the kscan_mchp_xec.c driver.)

ChiHuaL committed:
drivers: kscan: introduce new APIs for kscan matrix

Two types of keyboard scan drivers are supported in the codebase:
    - Remote type: the kscan hardware/algorithm exists in a remote chip.
      The SoC gets the key code and press/release event via a peripheral
       bus like I2C.
    - Embedded type: the kscan hardware/algorithm exists in the SoC.

For the second case, the existing kscan driver generally has two major
tasks to do:
    - control/receive the hardware signals, such as drive a column or
      read a row.
    - implement the kscan algorithm task.

However, an application may want to implement its scan algorithm by
driving the low-level hardware directly.
This commit extends the APIs to achieve it because the kscan matrix
needn't adhere to the driver anymore.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>
Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

ChiHuaL committed:
dts: kscan: npcx: add the kscan node and binding

* Add kscan matrix node to npcx.dtsi.
* Add yaml binding for kscan matrix driver.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>
Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

ChiHuaL committed:
driver: kscan: npcx: add driver support for kscan

* Add low-level keyboard scan driver support.
* Add kscan matrix scan driver support.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>
Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

ChiHuaL commented on dts/bindings/kscan/nuvoton,npcx-kscan.yaml:
```diff
@@ -0,0 +1,32 @@
+        required: true
+        description: configurations of pinmux controllers
+
+    wui_maps:
```
Done.

ChiHuaL commented on drivers/kscan/kscan_npcx.c:
```diff
@@ -0,0 +1,285 @@
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) ((const struct kscan_npcx_config *)(dev)->config)
+#define DRV_DATA(dev) ((struct kscan_npcx_data *)(dev)->data)
```
Done by using the style in #41918.

ChiHuaL commented on drivers/kscan/kscan_npcx.c:
```diff
@@ -0,0 +1,285 @@
+
+static struct kscan_npcx_data kscan_data;
+
+DEVICE_DT_INST_DEFINE(0, kscan_npcx_init, NULL, &kscan_data, &kscan_cfg, POST_KERNEL,
```
done.

ChiHuaL commented on include/drivers/kscan.h:
```diff
@@ -128,6 +161,88 @@ static inline int z_impl_kscan_disable_callback(const struct device *dev)
+ * @retval 0 If successful.
+ * @retval Negative errno code if failure.
+ */
+__syscall int kscan_config_matrix(const struct device *dev, struct kscan_matrix_config *config);
+
+static inline int z_impl_kscan_config_matrix(const struct device *dev,
+					     struct kscan_matrix_config *config)
```
The new API support for userspace is added. Thanks for pointing it out.

jackrosenthal commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,315 @@
+	for (int c = 0; c < CONFIG_KSCAN_MATRIX_MAX_COLUMNS; c++) {
+		deb_col = kb_poll.matrix_unstable_state[c];
+
+		if (!deb_col)
```
braces required

ChiHuaL replied:
done.

jackrosenthal commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,315 @@
+			uint8_t row_bit = matrix_new_state[c] & mask;
+
+			/* Continue if we already debounce a key */
+			if (!(deb_col & mask))
```
here too

ChiHuaL replied:
done

jackrosenthal commented on drivers/kscan/kscan_npcx.c:
```diff
@@ -0,0 +1,283 @@
+	if (resume)
+		irq_enable(config->irq);
+	else
+		irq_disable(config->irq);
```
braces required

ChiHuaL replied:
done.

keith-zephyr commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,323 @@
+			if (check_key_events(kscan_dev)) {
+				local_poll_timeout = kb_poll.poll_timeout;
+				start_poll_cycles = k_cycle_get_32();
+			} else if (!poll_expired(start_poll_cycles, &local_poll_timeout)) {
```
Does it make sense to rework this polling timeout to use `sys_clock_timeout_end_calc()`?  That [routine](https://docs.zephyrproject.org/latest/reference/kernel/timing/clocks.html#subsystems-using-k-timeout-t) seems purposely written to support this use case.

It would be good to eliminate `CLOCK_32K_HW_CYCLES_TO_US` if you can.

ChiHuaL replied:
The latest patch used sys_clock_timeout_end_calc for the polling timeout.
Also, I replaced the CLOCK_32K_HW_CYCLES_TO_US macro with Zephyr kernel API "k_cyc_to_us_floor32"

ChiHuaL committed:
boards: npcx_evb: add kscan DT node to npcx_evb

Add kscan device node to npcx9 and npcx7 EVB.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>
Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

ChiHuaL commented on include/drivers/kscan.h:
```diff
@@ -44,6 +50,25 @@ typedef void (*kscan_callback_t)(const struct device *dev, uint32_t row,
+ * This defines keyboard matrix configuration used by keyboard scan task and driver.
+ */
+struct kscan_matrix_config {
+	kscan_callback_t callback;
```
done

keith-zephyr replied:
I'm confused about this callback.  This callback is getting set by the `kscan_npcx.c` in it's `config_matrix` routine:

```
static int kscan_npcx_configure_matrix(const struct device *dev, struct kscan_matrix_config *config)
{
	struct kscan_npcx_data *const data = dev->data;

	data->config = config;

	/* Assign enable flag to task configuration */
	config->enable_scan = &data->enable_scan;
	/* Configure callback function between kscan task and driver */
	config->callback = kscan_npcx_matrix_callback;

	return 0;
}
```

So the kscan matrix polling task calls the NPCX driver, which then calls the user's callback.

Can the polling task call the user's callback directly?

Do you anticipate users/applications calling both `kscan_config()` and `kscan_matrix_config()`?  Seems like an application will only use one or the other.

ChiHuaL replied:
>Can the polling task call the user's callback directly?  

The user's callback is registered via the kscan_config(), the polling task has no information of the user's callback.
That why the polling task needs the member callback in the `struct kscan_matrix_config`  to link to kscan_npcx_matrix_callback which call the user's callback.

>Do you anticipate users/applications calling both kscan_config() and kscan_matrix_config()? Seems like an application will only use one or the other.

To test the kscan driver and kscan matrix polling task, we run the [kscan sample code](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/kscan) on NPCX EVB. The application (src/main.c) registers its callback to npcx kscan driver via kscan_config().
The polling task get a link to the user's callback via kscan_matrix_config().
After the polling task finishes the scanning and detect a key press/release event, it can use the callback to report the event to the  application.

keith-zephyr replied:
As the code works today, the user registers a callback with the kscan driver, and the driver directly invokes this callback after keypress events.

This PR is effectively adding a new keyboard type - matrix keyboards.

What about changing this up so that the `kscan_matrix` becomes it's own keyboard driver?  The NPCX driver would be a child of the matrix driver. 

This would work as follows:

The `kscan_npcx.c` driver would register itself with the `kscan_matrix` driver, using a new API `kscan_matrix_api`.  `kscan_matrix` then registers itself with the existing `kscan` driver.

This would look like:
```
/* kscan_npcx.c */
static const struct kscan_matrix_api kscan_matrix_ncpx_driver_api = {
	.config_matrix = kscan_npcx_configure_matrix,
	.drive_matrix_column = kscan_npcx_drive_matrix_column,
	.read_matrix_row = kscan_npcx_read_matrix_row,
	.resume_matrix_detection = kscan_npcx_resume_matrix_detection,
}

/* kscan_matrix.c */
static const struct kscan_driver_api kscan_matrix_driver_api = {
	.config = kscan_npcx_configure,
	.enable_callback = kscan_npcx_enable_interface,
	.disable_callback = kscan_npcx_disable_interface,
};
```

The users still call `kscan_config()`, `kscan_disable_callback()`, and `kscan_enable_callback()`.  But these routines call into the `kscan_matrix` driver, and the matrix driver then calls the chip level driver for the matrix operations.

It looks like the ITE and MCHP drivers could be converted to used the `kscan_matrix` driver.

The logical organization would be:

```
kscan drivers:
    cst8186s
    ft5336
    ht16k33
    matrix
        npcx
        ite
        mchip

```

ChiHuaL replied:
Done.

ChiHuaL commented on drivers/kscan/Kconfig:
```diff
@@ -26,4 +27,59 @@ config KSCAN_INIT_PRIORITY
+	bool
+	help
+	  "A matrix circuit is used in most computer keyboards where the key
+	  switches are connected by a grid of wires, similar to a diode matrix.
```
Done.

ChiHuaL commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,323 @@
+			 * flipped the bits when the matrix was scanned.
+			 * now we or the colums using z&(z-1) which is
+			 * non-zero only if z has more than one bit set.
+			 */
```
Done.

ChiHuaL commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,323 @@
+			continue;
+		}
+
+		for (int c_n = c + 1; c_n < CONFIG_KSCAN_MATRIX_MAX_COLUMNS; c_n++) {
```
done.

ChiHuaL commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,323 @@
+		kscan_drive_matrix_column(kscan_dev, col);
+
+		/* Allow the matrix to stabilize before reading it */
+		k_msleep(10);
```
done.

ChiHuaL commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,323 @@
+
+	/* Abort if ghosting is detected */
+	if (is_matrix_ghosting(matrix_new_state)) {
+		return false;
```
The code of the kscan matrix poll task is extracted from the [kscan_mchp_xec](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/kscan/kscan_mchp_xec.c#L249) driver and it return false.
I don't understand why it return false here.
Anyway, I have changed the task to return  any_pressed to meet the  logic of Chromium EC keyboard_scan.c because the embedded keyboard scan drivers are majorly used in the Chromebook application.

ChiHuaL commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,305 @@
+LOG_MODULE_REGISTER(kb_poll);
+
+/* Milliseconds in microseconds */
+#define MSEC_PER_MS 1000U
```
Done.

gmarull commented on include/drivers/kscan.h:
```diff
@@ -24,6 +24,12 @@
 extern "C" {
 #endif
 
+/* Drive all columns of keyboard matrix to low */
+#define KEYBOARD_COLUMN_DRIVE_ALL -2
+/* Tri-state all columns of keyboard matrix */
+#define KEYBOARD_COLUMN_DRIVE_NONE -1
```
please use Doxygen comments, `/**` and prefix macros with `KSCAN_`

ChiHuaL replied:
Done

gmarull commented on include/drivers/kscan.h:
```diff
@@ -24,6 +24,12 @@
+#define KEYBOARD_COLUMN_DRIVE_ALL -2
+/* Tri-state all columns of keyboard matrix */
+#define KEYBOARD_COLUMN_DRIVE_NONE -1
+/* 0 ~ MAX_MATRIX_KEY_COLS-1 Drive specified column of keyboard matrix to low */
```
?

ChiHuaL replied:
It is unnecessary comment and was removed.

gmarull commented on include/drivers/kscan.h:
```diff
@@ -128,6 +167,95 @@ static inline int z_impl_kscan_disable_callback(const struct device *dev)
+	const struct kscan_driver_api *api = (struct kscan_driver_api *)dev->api;
+
+	if (api->config_matrix == NULL) {
+		return -ENOTSUP;
```
`-ENOSYS`

gmarull replied:
same for others ^

ChiHuaL replied:
Done by replacing all ENPTSUP with -ENOSYS.

gmarull commented on include/drivers/kscan.h:
```diff
@@ -128,6 +167,95 @@ static inline int z_impl_kscan_disable_callback(const struct device *dev)
 }
 
+/**
+ * @brief Configure a Keyboard matrix scan instance.
```
Please document this is optional

ChiHuaL replied:
Done.

gmarull commented on include/drivers/kscan.h:
```diff
@@ -128,6 +167,95 @@ static inline int z_impl_kscan_disable_callback(const struct device *dev)
+ * event such as key pressed/released.
+ *
+ * @retval 0 If successful.
+ * @retval Negative errno code if failure.
```
`-ENOSYS` missing in this list

ChiHuaL replied:
Done.

gmarull commented on include/drivers/kscan.h:
```diff
@@ -44,6 +50,31 @@ typedef void (*kscan_callback_t)(const struct device *dev, uint32_t row,
+ * This defines keyboard matrix configuration used by keyboard scan task and driver.
+ */
+struct kscan_matrix_config {
+	/*
+	 * The callback is called when the keyboard matrix poll task detects a col and row
+	 * release/press event. It is registered when kscan_config_matrix API is called.
+	 * It is the link to the user/application callback which is registered when the
+	 * kscan_config API is called by the application.
+	 */
+	kscan_callback_t callback;
+	kscan_isr_callback_t isr_callback;
+	atomic_t *enable_scan;
```
so why do we need to types of callbacks? Where is this documented?

keith-zephyr replied:
I don't see a need for the `isr_callback`.  The only user is `kscan_matrix_poll.c`.  You could have the drivers  directly notify the matrix task.

```
static void kscan_npcx_ksi_isr(const struct device *dev, struct npcx_wui *wui)
{
	ARG_UNUSED(wui);

	if (IS_ENABLED(CONFIG_KCAN_MATRIX_TASK)) {
		kscan_matrix_scan_complete(dev);
	}
}
```

Alternately, you could pass a reference to the semaphore to the driver.

```
	if (data->config->matrix_task_sem) {
		k_sem_give(data->config->matrix_task_sem);
	}
```

gmarull commented on dts/bindings/kscan/nuvoton,npcx-kscan.yaml:
```diff
@@ -0,0 +1,32 @@
+    clocks:
+        required: true
+
+    pinctrl-0:
+        type: phandles
+        required: true
+        description: configurations of pinmux controllers
```
gentle reminder: a pinctrl driver has to be implemented, this will be deprecated soon. https://github.com/zephyrproject-rtos/zephyr/issues/39740

ChiHuaL replied:
Thanks for the reminder, Nuvoton has already provided an PR  proposal #43638 for the Pinctrl driver.
We will refactor all npcx drivers to use the Pinctrl API once the driver is accepted.

gmarull commented on drivers/kscan/Kconfig:
```diff
@@ -27,4 +28,66 @@ config KSCAN_INIT_PRIORITY
+
+if KSCAN_MATRIX
+
+config KSCAN_MATRIX_MAX_COLUMNS
+	int "keyboard matrix max column"
+	default 18
+	help
+	  Maximum columns of the matrix circuit.
+
+config KSCAN_MATRIX_MAX_ROWS
+	int "keyboard matrix max row"
+	default 8
+	help
+	  Maximum rows of the matrix circuit.
```
why isn't this part of Devicetree kscan devices? it could be included by kscan bindings, same it happens for width/height on displays.

ChiHuaL replied:
Nuvoton EC SoC can support maximum 18 columns x 8 rows for kscan matrix hardware interface.
However, in an application such as Chromebook, it usually uses 13X8 or 15x8 configuration and left the reminding pins for other usage. That's reason I added these two Kconfig entries. The application can overwrite the default values according to its requirement.

gmarull commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,303 @@
+{
+	ARG_UNUSED(dev);
+
+	kscan_dev = device_get_binding(KSCAN_DEV_NAME);
```
please use DEVICE_DT_GET, and check for readiness

gmarull commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,303 @@
+
+	kscan_config_matrix(kscan_dev, &kb_poll.config);
+
+	k_thread_create(&kb_poll.thread, kb_poll.thread_stack, TASK_STACK_SIZE, polling_task, NULL,
+			NULL, NULL, K_PRIO_COOP(4), 0, K_NO_WAIT);
```
why does this code think `K_PRIO_COOP(4)` will always be the right choice? shouldn't this be configurable?

ChiHuaL replied:
Done. Added a new Kconfig option "KSCAN_MATRIX_THREAD_PRIO" for the thread priority.

gmarull commented on drivers/kscan/kscan_npcx.c:
```diff
@@ -0,0 +1,286 @@
+	atomic_t enable_scan;
+};
+
+#define HAL_INSTANCE(dev)                                                                          \
+	((struct kbs_reg *)((const struct kscan_npcx_config *)(dev)->config)->base)
```
please avoid such kind of helpers, check what e.g. cleanup of serial drivers

ChiHuaL replied:
Done.

gmarull commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,303 @@
+	return 0;
+}
+
+SYS_INIT(kscan_matrix_poll_task_init, POST_KERNEL, CONFIG_KSCAN_MATRIX_TASK_INIT_PRIORITY);
```
this is not a driver, so I'm not sure its place is `drivers/kscan`. I wonder why this isn't provided as an API enhancement, a sort of non-device specific layer on top of the API.

gmarull reviewed (changes requested):
I think it's worth discussing this in the API meeting.

keith-zephyr commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,307 @@
+
+static void kscan_ksi_isr_callback(const struct device *dev)
+{
+	kscan_resume_matrix_detection(kscan_dev, false);
```
This statement is duplicated by `polling_task()` after it does `k_sem_take()`.

ChiHuaL replied:
>I don't see a need for the isr_callback. The only user is kscan_matrix_poll.c. You could have the drivers directly notify the matrix task.

This is intended to do it. 
I assumed that the kscan driver is unware about how the polling task react to the KSI's level change.
For example, this Zephyr's  ksan matrix polling task uses k_sem_give/k_sem_task mechanism to start/wait for the scanning.
The Cros EC's keyboard_scan_task uses the [task_wake(TASK_ID_KEYSCAN)](https://chromium-review.googlesource.com/c/chromiumos/platform/ec/+/3486597/6/zephyr/drivers/cros_kb_raw/cros_kb_raw.c#80) (i. e. k_polling_signal*) mechanism to do it.

>This statement is duplicated by polling_task() after it does k_sem_take().

Done by removing this statemeent.

keith-zephyr commented on drivers/kscan/kscan_handlers.c:
```diff
@@ -32,3 +32,35 @@ static int z_vrfy_kscan_enable_callback(const struct device *dev)
 }
 #include <syscalls/kscan_enable_callback_mrsh.c>
+
+static int z_vrfy_kscan_config_matrix(const struct device *dev, struct kscan_matrix_config *config)
```
Should all these APIs use `kscan_matrix` as the prefix?

```
kscan_matrix_config()
kscan_matrix_drive_column()
kscan_matrix_read_row()
kscan_matrix_resume_detection()
```

ChiHuaL replied:
Done.

keith-zephyr commented on include/drivers/kscan.h:
```diff
@@ -55,11 +89,19 @@ typedef int (*kscan_config_t)(const struct device *dev,
 	kscan_config_t config;
 	kscan_disable_callback_t disable_callback;
 	kscan_enable_callback_t enable_callback;
+	kscan_config_matrix_t config_matrix;
+	kscan_drive_matrix_column_t drive_matrix_column;
+	kscan_read_matrix_row_t read_matrix_row;
+	kscan_resume_matrix_detection_t resume_matrix_detection;
```
This should be guarded with CONFIG_KCAN_MATRIX.

ChiHuaL replied:
Done.

mbolivar-nordic reviewed (commented):
ACK on the devicetree parts

ChiHuaL committed:
sample: drivers: kscan: add support for npcx7/npcx9 EVB

This commit adds the necessary CONFIG* to run the kscan sample on
npcx7m6fb_evb and npcx9m6f_evb.

Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

ChiHuaL committed:
sample: driver: kscan: add support to use kscan matrix device

This commit adds support to run the sample for kscan matrix driver when
CONFIG_KSCAN_MATRIX_DRIVER is defined.

Signed-off-by: Jun Lin <CHLin56@nuvoton.com>

keith-zephyr commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,342 @@
+			 * application about the keys pressed.
+			 */
+			data->matrix_stable_state[c] ^= mask;
+			data->callback(dev, r, c, row_bit ? true : false);
```
Guard with `if (data->callback)`

ChiHuaL replied:
Done.

keith-zephyr commented on soc/arm/nuvoton_npcx/npcx7/Kconfig.defconfig.series:
```diff
@@ -50,6 +50,14 @@ config TACH_NPCX
 
+config KSCAN_NPCX
+	default y
+	depends on KSCAN
```
Delete - put the depends on the `KSCAN_MATRIX_NPCX` defintion.

ChiHuaL replied:
Done. Also delete this in soc/arm/nuvoton_npcx/npcx9/Kconfig.defconfig.series

keith-zephyr commented on drivers/kscan/Kconfig.npcx:
```diff
@@ -0,0 +1,20 @@
+
+config KSCAN_NPCX
+	bool "Nuvoton NPCX embedded controller (EC) Keyboard Scan (KSCAN) driver"
+	depends on SOC_FAMILY_NPCX
```
`depends on SOC_FAMILY_NPCX && KSCAN_MATRIX`

ChiHuaL replied:
Done.

keith-zephyr commented on drivers/kscan/kscan_npcx.c:
```diff
@@ -0,0 +1,230 @@
+/*
```
Rename file to `kscan_matrix_npcx.c`

ChiHuaL replied:
Done.

keith-zephyr commented on include/zephyr/drivers/kscan.h:
```diff
@@ -55,12 +72,27 @@ typedef int (*kscan_config_t)(const struct device *dev,
+
+#ifdef CONFIG_KSCAN_MATRIX
+__subsystem struct kscan_matrix_driver_api {
+	kscan_matrix_config_t config_matrix;
+	kscan_matrix_drive_column_t drive_matrix_column;
+	kscan_matrix_read_row_t read_matrix_row;
+	kscan_matrix_resume_detection_t resume_matrix_detection;
```
Make the naming consistent:
```
	kscan_matrix_config_t matrix_config;
	kscan_matrix_drive_column_t matrix_drive_column;
	kscan_matrix_read_row_t matrix_read_row;
	kscan_matrix_resume_detection_t matrix_resume_detection;
```

ChiHuaL replied:
Done.

keith-zephyr commented on drivers/kscan/kscan_npcx.c:
```diff
@@ -0,0 +1,230 @@
+}
+
+static const struct kscan_matrix_driver_api kscan_matrix_npcx_driver_api = {
+	.config_matrix = kscan_npcx_configure_matrix,
+	.drive_matrix_column = kscan_npcx_drive_matrix_column,
+	.read_matrix_row = kscan_npcx_read_matrix_row,
+	.resume_matrix_detection = kscan_npcx_resume_matrix_detection,
```
Update the naming of the local functions to match kscan_matrix_driver_api.
```
	.matrix_config = kscan_matrix_npcx_config,
	.matrix_drive_column = kscan_matrix_npcx_drive_column,
	.matrix_read_row = kscan_matrix_npcx_read_row,
	.matrix_resume_detection = kscan_matrix_npcx_resume_detection,```

ChiHuaL replied:
Done.

keith-zephyr commented on soc/arm/nuvoton_npcx/npcx9/Kconfig.defconfig.series:
```diff
@@ -50,6 +50,14 @@ config TACH_NPCX
 
+config KSCAN_NPCX
+	default y
+	depends on KSCAN
```
Delete - use the depends in kscan/Kconfig.npcx

ChiHuaL replied:
Done.

keith-zephyr commented on drivers/kscan/Kconfig:
```diff
@@ -27,4 +27,77 @@ config KSCAN_INIT_PRIORITY
+	default 95
+	help
+	  Keyboard matrix task initialization priority.
+config KSCAN_MATRIX_THREAD_PRIO
```
Add a space before this config.

ChiHuaL replied:
Done.

keith-zephyr commented on drivers/kscan/Kconfig:
```diff
@@ -27,4 +27,77 @@ config KSCAN_INIT_PRIORITY
+	help
+	  Maximum rows of the matrix circuit.
+
+config KSCAN_MATRIX_DRIVER
```
Move the KSCAN_MATRIX_DRIVER configs into a separate file.

ChiHuaL replied:
I moved the whole KSCAN_MATRIX configs into Kconfig.matrix file.

keith-zephyr commented on drivers/kscan/Kconfig:
```diff
@@ -27,4 +27,77 @@ config KSCAN_INIT_PRIORITY
+endif # KSCAN_MATRIX_DRIVER
+
+comment "Keyboard Scan Matirx Chip Drivers"
+source "drivers/kscan/Kconfig.npcx"
```
Move this with the other drivers up above.  Note that `if/endif` blocks in Kconfig files do not skip `source` statements.

https://docs.zephyrproject.org/latest/build/kconfig/tips.html#lack-of-conditional-includes

ChiHuaL replied:
Thanks for the reminder.
The original thinking to put source source "drivers/kscan/Kconfig.npcx" here is let the layout of npcx selection within the KSCAN_MATRIX scope.
Anyway, I have moved this line up, together with the other  source "xxx files".

keith-zephyr commented on samples/drivers/kscan/src/main.c:
```diff
@@ -15,9 +15,13 @@
 
 LOG_MODULE_REGISTER(main);
 
+#ifdef CONFIG_KSCAN_MATRIX_DRIVER
```
If you create dts/bindings/kscan/kscan-matrix.yaml for the KSCAN_MATRIX_DRIVER, then you won't need to change this sample.

ChiHuaL replied:
Considering that the kscan_matrix.c is a pseudo device, I am not sure where is a propriate DTS location to put this device node. That's the reason I created the device name in the Kconfig file.

keith-zephyr replied:
I see there is precedence for this approach from these devices:
`CONFIG_ETH_NATIVE_POSIX_DRV_NAME`
`CONFIG_CRYPTO_TINYCRYPT_SHIM_DRV_NAME`
`CRYPTO_MBEDTLS_SHIM_DRV_NAME`

keith-zephyr reviewed (changes requested):
Thanks for refactoring.  I think this is a much better logical organization and I found the code easier to follow.  Especially the callback path.

ChiHuaL commented on drivers/kscan/kscan_handlers.c:
```diff
@@ -32,3 +32,39 @@ static int z_vrfy_kscan_enable_callback(const struct device *dev)
 }
 #include <syscalls/kscan_enable_callback_mrsh.c>
+
+#ifdef CONFIG_KSCAN_MATRIX
```
Done.

ChiHuaL commented on include/zephyr/drivers/kscan.h:
```diff
@@ -24,6 +24,15 @@
+/**
+ * @brief Drive all columns of keyboard matrix to low.
+ */
+#define KEYBOARD_COLUMN_DRIVE_ALL -2
```
Done.

ChiHuaL commented on dts/bindings/kscan/nuvoton,npcx-kscan.yaml:
```diff
@@ -0,0 +1,32 @@
+
+description: Nuvoton, NPCX keyboard matrix controller
+
+compatible: "nuvoton,npcx-kscan"
```
Done.

ChiHuaL commented on drivers/kscan/Kconfig.npcx:
```diff
@@ -0,0 +1,20 @@
+	help
+	  This option enables the KSCAN driver for NPCX family of
+	  processors.
+	  Say y if you wish to use keyboard scan driver on NPCX MCU.
```
Done.

ChiHuaL commented on drivers/kscan/Kconfig.npcx:
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2022 Nuvoton Technology Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+config KSCAN_NPCX
```
Done.

ChiHuaL commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,342 @@
+/*
```
Both are done.

ChiHuaL commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,342 @@
+	return key_pressed;
+}
+
+void polling_task(const struct device *dev, void *dummy2, void *dummy3)
```
Done.

ChiHuaL commented on drivers/kscan/kscan_matrix_poll.c:
```diff
@@ -0,0 +1,342 @@
+/* Number of tracked scan times */
+#define SCAN_OCURRENCES 30U
+/* Thread stack size */
+#define TASK_STACK_SIZE 1024
```
Done.

keith-zephyr commented on drivers/kscan/Kconfig.matrix:
```diff
@@ -0,0 +1,83 @@
+
+if KSCAN_MATRIX_DRIVER
+
+config KSCAN_MATRIX_DRV_NAME
```
Please add some help text for this option.

ChiHuaL replied:
Done.

keith-zephyr reviewed (dismissed):
@mbolivar-nordic can you take another look?

gmarull commented on include/zephyr/drivers/kscan_matrix.h:
```diff
@@ -0,0 +1,166 @@
+/*
```
why isn't this _new_ API part of `kscan.h`?

ChiHuaL replied:
It is requested during the [review conversation](https://github.com/zephyrproject-rtos/zephyr/pull/42923#discussion_r849667743).

gmarull commented on include/zephyr/drivers/kscan_matrix.h:
```diff
@@ -0,0 +1,166 @@
+extern "C" {
+#endif
+
+/**
+ * @brief Drive all columns of keyboard matrix to low.
+ */
+#define KEYBOARD_COLUMN_DRIVE_ALL -2
+/**
+ * @brief Tri-state all columns of keyboard matrix.
+ */
+#define KEYBOARD_COLUMN_DRIVE_NONE -1
```
enum?

ChiHuaL replied:
Done.

gmarull commented on include/zephyr/drivers/kscan_matrix.h:
```diff
@@ -0,0 +1,166 @@
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ */
+typedef void (*kscan_isr_callback_t)(const struct device *dev);
```
One of the problems I see with this API is that we now have `kscan_isr_callback_t` and `kscan_callback_t`, which is confusing at best (more counting this API is added with no documentation at all).

gmarull commented on drivers/kscan/Kconfig.matrix:
```diff
@@ -0,0 +1,84 @@
+
+if KSCAN_MATRIX
+
+config KSCAN_MATRIX_MAX_COLUMNS
+	int "keyboard matrix max column"
+	default 18
+	help
+	  Maximum columns of the matrix circuit.
+
+config KSCAN_MATRIX_MAX_ROWS
+	int "keyboard matrix max row"
+	default 8
+	help
+	  Maximum rows of the matrix circuit.
```
why isn't this described in Devicetree? What if there are 2 devices with different configs?

ChiHuaL replied:
This follows the existing drivers which support the keyboard matrix hardware in the SoC.
As the keyboard matrix requires a lot of pads for column/row, most EC chips support only one hardware module from a cost perspective.

gmarull commented on drivers/kscan/kscan_matrix.c:
```diff
@@ -0,0 +1,353 @@
+ * general kscan driver using the kscan_driver_api.
+ */
+
+#include <errno.h>
+#include <device.h>
+#include <drivers/kscan.h>
+#include <drivers/kscan_matrix.h>
+#include <kernel.h>
+#include <soc.h>
+#include <logging/log.h>
```
all headers needs `<zephyr/>` prefix

ChiHuaL replied:
Done.

gmarull commented on drivers/kscan/kscan_matrix.c:
```diff
@@ -0,0 +1,353 @@
+	.enable_callback = kscan_matrix_enable_interface,
+	.disable_callback = kscan_matrix_disable_interface,
+};
+DEVICE_DEFINE(kscan_matrix, CONFIG_KSCAN_MATRIX_DRV_NAME, kscan_matrix_init, NULL, &kb_matrix_data,
```
why not DT based devices?

ChiHuaL replied:
The kscan_matrix is a pseudo driver that registers with the existing kscan interface. It doesn't require bindings to the actual hardware peripheral.

gmarull commented on drivers/kscan/kscan_matrix_npcx.c:
```diff
@@ -0,0 +1,231 @@
+		kscan_npcx_init_ksi_wui_callback(dev, &ksi_callback[i], &config->wui_maps[i],
+						 kscan_npcx_ksi_isr);
+
+	/* Configure pin-mux for kscan device */
+	npcx_pinctrl_mux_configure(config->alts_list, config->alts_size, 1);
```
please use pinctrl API

ChiHuaL replied:
Done.

gmarull reviewed (changes requested):
This addition feels like an ad-hoc solution for one usecase, I think it needs some documentation to explain its design or how it can be used. It could also be presented in the API meeting.

keith-zephyr commented:
> This addition feels like an ad-hoc solution for one usecase, I think it needs some documentation to explain its design or how it can be used. It could also be presented in the API meeting.

In addition to the Nuvoton keyboard driver, the matrix style keyboard support can be shared by used by the [ITE](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/kscan/kscan_ite_it8xxx2.c) and [Microchip](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/kscan/kscan_mchp_xec.c) keyboard drivers.

Essentially the kscan_matrix is a virtual driver than registers with the existing kscan interface.  The Nuvoton, ITE, and Microchip keyboard drivers register with the kscan_matrix virtual driver.

mbolivar-nordic commented on drivers/kscan/kscan_matrix_handlers.c:
```diff
@@ -0,0 +1,41 @@
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_KSCAN(dev, matrix_read_row));
+
+	return z_impl_kscan_matrix_read_row((const struct device *)dev, row);
```
Are you sure about this? It looks to me like a malicious userspace thread could pass an address in `row` that they don't have write access to. But the (privileged mode) driver unconditionally writes to the address pointed by `row` here:

```c
	/* 1 means key pressed, otherwise means key released. */
	*row = (~val & NPCX_KB_ROW_MASK);
```

... this looks like a perfect recipe for an arbitrary write primitive. What am I missing here?

dcpleung replied:
In addition to verifying `*dev`, you will also need to verify that `*row` can be readable, or writable, or read/write-able by using `Z_SYSCALL_MEMORY_*`.

ChiHuaL replied:
I added the write permission check for it:
`Z_OOPS(Z_SYSCALL_MEMORY_WRITE(row, sizeof(*row))); `

mbolivar-nordic commented on drivers/kscan/kscan_matrix_handlers.c:
```diff
@@ -0,0 +1,41 @@
+{
+	Z_OOPS(Z_SYSCALL_DRIVER_KSCAN(dev, matrix_config));
+
+	return z_impl_kscan_matrix_configure((const struct device *)dev, isr_callback);
```
This similarly looks like a security vulnerability to me. The `isr_callback` is a function pointer that is attacker-controlled, and is saved by the implementation to be invoked from (privileged mode) ISR context by the driver here:

```
	if (data->isr_callback != NULL) {
		data->isr_callback(dev);
	}
```

This similarly seems to allow the thread to escape its sandbox.

dcpleung replied:
ISR callback registration for drivers should never be done from user mode.

mbolivar-nordic replied:
Thanks for confirming both issues @dcp

ChiHuaL replied:
I removed the callback function register API from the kscan_matrix_handler.

mbolivar-nordic reviewed (dismissed):
I am generally concerned about the review quality on the userspace handlers. The current implementations seem to have two obvious flaws to me. Am I just missing something here?

Though @dcpleung is the userspace maintainer, I don't think we can ask him to review all the handlers coming into the tree, but I'm concerned that we're otherwise not taking the requirements of user mode's threat model (for docs, see: https://docs.zephyrproject.org/latest/kernel/usermode/overview.html) seriously enough here.

I am going to keep my -1 on here unless one of these 3 things happens:

1. someone can explain what I am wrong about in the following review comments
2. the bugs are fixed
3. userspace support for this API is removed

I didn't look in detail at @gmarull's review, but some of his comments seem like they have some merit as well.

mbolivar-nordic commented:
@dcpleung would you mind taking a look at the updated handlers?

dcpleung commented:
`z_vrfy_kscan_matrix_read_row` looks good.
